src/main/java/dk/cngroup/trainings/spring/springassignment/service/AnimalServiceImpl.java

```diff
@@ -1,80 +1,109 @@
-    }
+	public AnimalServiceImpl(AnimalRepository animalRepository, CareTakerRepository careTakerRepository) {
+		this.animalRepository = animalRepository;
+		this.careTakerRepository = careTakerRepository;
+	}
 
-    @Override
-    public List<Animal> getAnimals() {
-        return animalRepository.findAll();
-    }
+	@Override
+	public List<Animal> getAnimals() {
+		return animalRepository.findAll();
+	}
 
-    @Override
-    public Optional<Animal> getAnimalById(long id) {
-            return animalRepository.findById(id);
-    }
+	@Override
+	public Optional<Animal> getAnimalById(long id) {
+		return animalRepository.findById(id);
+	}
 
-    @Override
-    public List<Animal> getAnimalsByName(String name) {
-            return animalRepository.findAllByName(name);
-    }
+	@Override
+	public List<Animal> getAnimalsByName(String name) {
+		return animalRepository.findAllByName(name);
+	}
 
-    @Override
-    public Optional<Animal> addAnimal(Animal animal){
-        if(isValid(animal)){
-            if (!animalRepository.existsById(animal.getId())) {
-                return Optional.ofNullable(animalRepository.save(animal));
-            }else {
-                // Call addAnimal recursively with new id untill it's saved
-                animal.setId(animal.getId() + 1); //TODO check with generated value
-                return this.addAnimal(animal);
-            }
-        }
-        else {
-            return null; //TODO dont return null
-        }
-    }
+	@Override
+	public Optional<Animal> addAnimal(Animal animal) {
+		if (AnimalValidationService.isValid(animal)) {
+			animal.setId(IdGenerator.getId());
+			return Optional.ofNullable(animalRepository.save(animal));
+		} else {
+			return Optional.empty();
+		}
+	}
 
-    @Override
-    public boolean deleteAnimalById(long id) {
-        if(animalRepository.existsById(id)) {
-            animalRepository.deleteById(id);
-            return true;
-        }else
+	@Override
+	public boolean deleteAnimalById(long id) {
+		if (animalRepository.existsById(id)) {
+			animalRepository.deleteById(id);
+			return true;
+		} else
 
-            return false;
-    }
+		{
+			return false;
+		}
+	}
 
-    @Override
-    public Optional<Animal> updateAnimalById(long id, @Valid Animal animal) { //TODO delete @Valid
-        if(isValid(animal) && animalRepository.existsById(id)) {
-            // Change the updated animal id and override it in the database.
-            animal.setId(id);
-            return Optional.ofNullable(animalRepository.save(animal));
-        } else {
-            return null; //TODO Optional.empty();
-        }
-    }
+	@Override
+	public Optional<Animal> updateAnimalById(long id, Animal animal) {
+		if (AnimalValidationService.isValid(animal) && animalRepository.existsById(id)) {
+			// Change the updated animal id and override it in the database.
+			animal.setId(id);
+			return Optional.ofNullable(animalRepository.save(animal));
+		} else {
+			return Optional.empty();
```

Comment:
here you cannot distinquish between case when it was invalid animal and when the animal was not found. So maybe it would be better to throw some Exceptions here?

Reply:
I can refactor it to throwing InvalidAnimalException